text/0031-configuration.md

```diff
@@ -0,0 +1,280 @@
++ Feature name: `configuration`
++ Start date: 2020-05-06
++ RFC PR: [iotaledger/bee-rfcs#31](https://github.com/iotaledger/bee-rfcs/pull/31)
++ Bee issue: [iotaledger/bee#71](https://github.com/iotaledger/bee/issues/71)
+
+# Summary
+
+This RFC proposes a configuration pattern for Bee binary and library crates.
+
+# Motivation
+
+This RFC contains a set of recommendations regarding configuration management in order to ensure consistency across the
+different Bee crates.
+
+# Detailed design
+
+## Libraries
+
+- [serde](https://github.com/serde-rs/serde): the go-to framework for **ser**ializing and **de**serializing Rust data
+  structures efficiently and generically;
+- [toml-rs](https://github.com/alexcrichton/toml-rs): a TOML encoding/decoding library for Rust with
+  serialization/deserialization on top of `serde`;
+
+## Recommendations
+
+With the following recommendations, one can:
+- read a configuration builder from a file;
+- write a configuration to a file;
+- manually override a configuration builder;
+- construct the configuration;
+
+### Configuration Builder
+
+The [Builder](https://rust-lang.github.io/api-guidelines/type-safety.html#c-builder) pattern is very common in Rust when
+it comes to constructing complex objects like configurations.
+
+A configuration builder type should:
+- be named `*ConfigBuilder`;
+- derive the following traits;
+    - `Default` to easily implement the `new` method as convention;
+    - `Deserialize`, from `serde`, to deserialize from a configuration file;
+- provide a `new` method;
+- have `Option` fields;
+    - if there is a configuration file, the fields should have the same names as the keys;
+- provide setters to set/override the fields;
+- provide a `build` method constructing the actual configuration object;
+- have default values defined as `const` and set with `Option::unwrap_or`;
+
+Here is a small example fitting all these requirements.
+
+`config.toml`
+```toml
+[snapshot]
+meta_file_path  = "./data/snapshot/mainnet.snapshot.meta"
+state_file_path = "./data/snapshot/mainnet.snapshot.state"
+```
+
+`config.rs`
+```rust
+const DEFAULT_META_FILE_PATH: &str = "./data/snapshot/mainnet.snapshot.meta";
+const DEFAULT_STATE_FILE_PATH: &str = "./data/snapshot/mainnet.snapshot.state";
+
+#[derive(Default, Deserialize)]
```

Comment:
Shouldn't we return `SnapshotConfig` instead of `SnapshotConfigBuilder` when reading the configuration file from disk? Therefore move `Deserialize` to `SnapshotConfig`.

Reply:
No because you may want to override everything, for example with CLI arguments